### ISSUE TITLE
Post-publication maintenance

### ISSUE THREAD

**Issue by khinsen** [ReScience/ReScience#43]
I have tried to reproduce the results of some of the papers published in ReScience, with the goal of using them as examples in a course on reproducible research. The idea is to let students re-run the code and compare the outcomes, without necessarily understanding much about the contents - the goal of the exercise is to show how computational work can be published reproducibly, and what reproduction actually involves with today's state of the art. But before telling students what to do, I prefer to try it myself.

So far I haven't fully succeeded with any of the three papers I have tried, which makes me wonder if we should encourage authors to provide some level of post-publication maintenance of their submissions. Technically this would be easy, as anyone can open an issue on a repository in ReScience-Archives to report a problem.

Any opinions?

**Comment by labarba** [comment]
Seriously? This is both interesting in its own right, and distressing. As you know, I published a paper about our own work to replicate a previous study (from our group, though the student had graduated) that had been done "reproducibly." It was a huge challenge! In that case, we were using other code bases to seek replication of the findings. Yet we also used the same code (same version), and ran into difficulties.

I don't think the solution is post-publication "maintenance and support," though. I think what we need is to work harder to establish the standards of reproducible research. 

Certainly, we should allow readers to comment post-publication if they tried to reproduce the ReScience paper, and encountered difficulties. But we don't have a way of enforcing a level of response from the authors.

But isn't checking that the ReScience submission is really reproducible part of the peer-review process?

Great idea to have students re-execute "reproducible" articles, and se what happens!

**Comment by rougier** [comment]
That's a big surprise because reviewers do test for the result and this means running the original code. When you say you haven't fully succeeded, you mean you cannot run the code or you don't get the same result ?

@pdebuyl suggested in #42 to have minimal information on the different versions needed to run the replication, that might help for future submissions. However, I don't expect people to be able to run the code indefinitely because we know that at some point, some libraries won't be available or won't be maintained in the future. A replication is merely a way to say the advertised results are correct according to the paper original description (because they have been replicated).

Bu we should try hard to not have  "non-reproducible replication".

**Comment by Chris1221** [comment]
I realize that this may be a little too demanding of authors, but the issue could potentially be addressed through a continuous integration service with unit tests for major results. Of course this might not be possible for some things such as figures, but for major numerical results (i.e. the coefficient was found to be X) the output of said results could be tested at each `push` to ensure that changing other bits of the code do not change this finding. It would also ensure that at every revision the replication maintains its validity while simultaneously ensuring that the code is water tight so that the tests may be run in the first place. The coverage of the tests could also become part of the review process.

**Comment by khinsen** [comment]
Some more details so that you can appreciate the situation.

In Detorakis-2016, the list of dependencies is incomplete. The code requires scikit-image, which is not mentioned. This could easily be missed by a reviewer who already has scikit-image in his/her Python installation. Another problem is that the script `run_all.py` works only if the Python installation used for execution is the one accessible as `python` on the shell's search path. Again this wouldn't bother a reviewer whose installation conforms to the assumption.

In Maksimov-Albada-Diesmann-2016, the installation instructions lead to a compilation error when applied under MacOS X. I haven't explored this further as I am rather allergic to C++ code. The instructions are for Ubuntu, so this is perhaps to be expected. But if someone proposes a fix for other platforms, there should be a way to integrate this.

What I find the most worrying is Vitay-2016. Getting the code to run is no problem, but the figures I obtain differ from those in the paper. Not knowing anything about the science, I cannot say how significant the deviations between the curves are, but in some plots I see one curve completely missing, which looks suspicious.

**Comment by ctb** [comment]
This discussion is remarkably similar to issues sparked in my brain on
Monday's Scientific Paper of the Future workshop at Caltech.  Konrad's
issue arrived in my inbox at almost exactly the same time I was putting
together a blog post that touches on many of these same issues --

http://ivory.idyll.org/blog/2017-pof-software-archivability.html

Feedback welcome!

The continuous analysis idea is a good 'un, see Beaulieu-Jones and Greene
http://biorxiv.org/content/early/2016/06/01/056473

**Comment by khinsen** [comment]
@ctb Yes, your blog post is very appropriate. But it's not just time that is a problem. The code I am trying to use has been found OK by reviewers a few months ago. I haven't updated my scientific software since then. The problem is that the code relies on a specific platform that we cannot describe in any useful and verifiable way to the person wanting to do the replication.

Adding versioning information as @pdebuyl has suggested is one step, but we'd really need some automatic analysis of the authors' software installation that produces a machine-readable output - and of course the tool that handled these outputs may never be modified...

**Comment by Chris1221** [comment]
@ctb Interesting post and article, thanks for the links. It seems like lack of incentive may also be a contributing factor to people not taking up these tools. Would there be any value to an option "reproducibly generated outputs" section of a rescience article? The authors could provide the CI tests and links to generate specific key outputs that show their reproduction is accurate along with the coverage of their analysis in general. It would show that these key findings are held constant across all versions and a permanent (DOI?) link to the travis (or similar) for the repository which tested these findings could be included in the publication. I'm not sure if thats feasible though.

**Comment by khinsen** [comment]
@Chris1221 The main problem I see with any CI-based approach is: what happens when the tests fail? Who does the work of analyzing what went wrong and how to fix it? And who pays for it?

**Comment by Chris1221** [comment]
@khinsen Good points that I'm not sure how to answer. My initial reaction would be to say that should authors choose to include such tests then they must all pass for submission, review, and publication, and they would be required to maintain their code for a fixed period of time following publication; or, and this may be slightly unreasonable, to set up a docker container with the correct versions of all software used and use this for the testing. That should never fail even if, say, python updates or a dependency goes out of date. However, I would have no idea how to go about enforcing these guidelines, and think they may be slightly too onerous. I also am not sure how the journal should indicate that tests have failed if they do so post publication. Definitely some points for discussion if people are interested in pursing CI.

**Comment by StuntsPT** [comment]
I suppose that if someone is pushing something to the repository that is being "tested", then there is interest in at least part of the analyses that comprises it. Whoever is doing those pushes *should*, at least in theory be interested in understanding what is failing.
Regardless, I think a static docker container might be the answer here, and despite the effort required to create one, I think it might be worth it in this case.
Perhaps we could even consider having a couple of "pre-built" containers for several popular languages that can lower the entry barrier for someone willing to "dockerize" their research for the sake of reproducibility.

**Comment by khinsen** [comment]
An idea to reduce platform dependencies in the future: assign a third reviewer with a different scientific background, whose sole task is to re-run the computations from the instructions. This would happen after our current reviewing process.

The rationale is that someone who does not work in the same field as the authors is also more likely to use a different computing environment and share less tacit knowledge with the authors.

**Comment by oliviaguest** [comment]
Sure... but why a scientist? Why not a software engineer?

**Comment by pdebuyl** [comment]
I would ask why "after the review"? Is it to check that the final instructions are good enough?

The idea, if I understand, is to check whether one can go from a "bare OS" to "obtaining the results". For instance, I am a regular OS X user and if there are *implied assumptions* from the code authors, such as "have XCode", "have NumPy", "have root access", etc. they would be unmasked from this test.

So, having a procedure as follows:
- Open a brand new Windows/Mac/Linux machine
- Install the dev tools
- run the code

would actually have the same effect, even if performed by an existing referee, an editor, or an extra referee.

**Comment by oliviaguest** [comment]
So run everything in a container?

**Comment by pdebuyl** [comment]
I don't know very much about containers :-) My idea of this is

- insert a USB boot stick in the machine
- format
- proceed

but it is probably more time consuming than a container :-)

What about just using a stock OS in a VM?

**Comment by oliviaguest** [comment]
That's essentially what a container is, but better.  😉 You would prefer it for the above. See here: https://www.docker.com/what-docker

**Comment by pdebuyl** [comment]
Thanks, I'll have a look.

**Comment by Chris1221** [comment]
If this gains momentum and there is a push to dockerize the submission process, I would very much enjoy helping out. I'm sure that decision would have to be made by the editors together, but it would be something I would love to contribute to.

**Comment by khinsen** [comment]
@oliviaguest It would be great to have research software engineers for doing this, but that's still a rare species. I'd say we take whoever is willing (and able) to do the job.

@pdebuyl There are many ways to package up an execution environment. The choice of containers, VMs, or yet something else is not the primary one for me. The very first question is: should we expect all our authors to provide a packaged environment? This might well reduce the number of submissions we get. We'd also have to worry about storage of large binary files.

That said, I'll admit that I am not a big fan of Docker. Both for practical reasons (it adds bugs, complexity, and an instability over time of its own) and for theoretical ones (it's a workaround rather than a solution of the fundamental packaging problem, as I tried to explain in a [blog post](http://blog.khinsen.net/posts/2016/03/04/composition-is-the-root-of-all-evil/)). I see Docker as a tool for facilitating deployment, but only over short time spans.

**Comment by pdebuyl** [comment]
@khinsen I agree totally that we should not impose any particular platform. Computational science is made on many platforms, with many languages, etc.

Actually, what I had in mind is that the reviewing procedure would include running the code in a "clean" environment to debug environment issues. See this very timely blog post by @mikecroucher http://www.walkingrandomly.com/?p=6275

**Comment by ThomasA** [comment]
@khinsen inspired by https://github.com/ReScience/ReScience-submission/pull/28#issuecomment-296088109, I have some experience setting Travis CI up for a repository. Since each article has a separate repository under https://github.com/ReScience-Archives/, this should make it quite straight-forward to set up a script to run each paper's code. At least for Python, I have experience setting it up based on [Anaconda's miniconda](https://conda.io/miniconda.html).

**Comment by khinsen** [comment]
@ThomasA Do you think it's possible to provide a reasonable-length instruction page, with code templates if necessary/useful, to help potential authors use Travis CI for their ReScience submission? Even if it's only for Python, that would be a big step forward, considering that most ReScience papers use Python.

Beyond the immediate benefit for ReScience, this would have the pedagogical side-effect of familiarizing more people with continuous integration.

**Comment by ThomasA** [comment]
@khinsen I will have to think more about the practicalities. In particular, I imagine it is something that should ideally work from the point of submission and I am not quite sure how to best do that. The point is that all previous and current submissions live in the ReScience-submissions repo as pull requests and I do not know yet if one repo can have more than one individual Travis set-up. Having something that re-runs all current and previous submissions on each new submission or change to one under review will be much too redundant and too huge a job to run regularly.
Another potential problem is changing the Travis configuration for a replication once it moves to ReScience-Archives.
If the CI set-up is only expected to work for the final, ReScience-Archives version, I think it makes everything much easier, but that way we will not have the opportunity to test the CI set-up during review or to utilise the CI set-up to check that the replication can run at submission.
I'll get back when I know more. Any CI set-up experts are of course welcome to chime in...

**Comment by ThomasA** [comment]
This could be helpful https://lord.io/blog/2014/travis-multiple-subdirs/
